lib/connection.js

```diff
@@ -294,56 +315,47 @@ class Connection extends EventEmitter {
   // Close the connection
   //
   close() {
+    this._closed = true;
     this.stopHeartbeat();
     this.transport.end();
   }
 
-  // Set a timeout using timers.enroll()
-  //   milliseconds - amount of milliseconds
-  //   callback - callback function
-  //
-  setTimeout(milliseconds, callback) {
-    timers.enroll(this, milliseconds);
-    timers._unrefActive(this);
-    this.once('_timeout', callback);
-  }
-
-  // Clear a timeout set with Connection#setTimeout
-  //   handler - timer callback to remove
-  //
-  clearTimeout(handler) {
-    timers.unenroll(this);
-    this.removeListener('_timeout', handler);
-  }
-
   // Returns underlying transport
   //
   getTransport() {
     return this.transport.getRawTransport();
   }
 
-  // timers.enroll() timeout handler
+  // Prepare a JSTP message to be sent over this connection
+  //   message - a message to prepare
   //
-  _onTimeout() {
-    this.emit('_timeout');
+  _prepareMessage(message) {
+    return serde.stringify(message);
+  }
+
+  _sendWithBuffering(message, messageId) {
+    const preparedMessage = this._prepareMessage(message);
+    this._send(preparedMessage, message);
+    this.session._bufferMessage(messageId, preparedMessage);
   }
 
   // Send a JSTP message over this connection
-  //   message - a message to send
+  //   preparedMessage - a prepared message to send
+  //   message - a message object (used for development events thus optional)
   //
-  _send(message) {
-    const data = serde.stringify(message);
-    this.transport.send(data);
+  _send(preparedMessage, message) {
+    this.transport.send(preparedMessage);
 
     if (process.env.NODE_ENV !== 'production') {
-      this.emit('outgoingMessage', message);
+      this.emit('outgoingMessage', message || serde.parse(preparedMessage));
```

Comment:
`serde.parse()` is used directly here, but `serde.stringify()` is abstracted away into `_prepareMessage()`, which is a bit inconsistent.